In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import zipfile
import os
# Перша частина завдання - обробка датасету data_2d
# Шлях до zip-файлу
zip_path = "C:\\Users\\Igor\\Downloads\\ml-100k.zip"
extract_folder = '/new/ml-100k'

# Розпакування файл
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_folder)

# Список файлів у розпакованій папці
extracted_files = os.listdir(extract_folder)
extracted_files[0]

'ml-100k'

In [2]:
# Перерахування файлів в папці "data".
data_folder_path = os.path.join(extract_folder, extracted_files[0])
data_folder_path

'/new/ml-100k\\ml-100k'

In [3]:
data_files = os.listdir(data_folder_path)
data_files

['allbut.pl',
 'mku.sh',
 'README',
 'u.data',
 'u.genre',
 'u.info',
 'u.item',
 'u.occupation',
 'u.user',
 'u1.base',
 'u1.test',
 'u2.base',
 'u2.test',
 'u3.base',
 'u3.test',
 'u4.base',
 'u4.test',
 'u5.base',
 'u5.test',
 'ua.base',
 'ua.test',
 'ub.base',
 'ub.test']

In [4]:
!pip install scikit-surprise

In [4]:
import pandas as pd
from surprise import Dataset, Reader, SVD
from surprise.model_selection import cross_validate, GridSearchCV

# Завантажимо датасет movielens
data = Dataset.load_builtin('ml-100k')
data

In [1]:
from surprise import Dataset, Reader
from surprise.model_selection import cross_validate
from surprise.prediction_algorithms import SVD, SVDpp, NMF


data = Dataset.load_builtin('ml-100k')


reader = Reader(rating_scale=(1, 5))

# Побудова моделі SVD та проведення крос-валідації для підбору параметрів
svd_model = SVD()
svd_cv_results = cross_validate(svd_model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

# Побудова моделі SVD++  (SVD з використанням неявних оцінок) та проведення крос-валідації
svdpp_model = SVDpp()
svdpp_cv_results = cross_validate(svdpp_model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

# Побудова моделі NMF  (Non-negative Matrix Factorization) та проведення крос-валідації
nmf_model = NMF()
nmf_cv_results = cross_validate(nmf_model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
# Результати крос-валідації (похибки RMSE і MAE)
# Виведення результатів крос-валідації зберігаються для кожної моделі в відповідних змінних
# svd_cv_results, svdpp_cv_results та nmf_cv_results
print("SVD Results:")
print(svd_cv_results)
print("\nSVD++ Results:")
print(svdpp_cv_results)
print("\nNMF Results:")
print(nmf_cv_results)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9438  0.9250  0.9321  0.9380  0.9327  0.9343  0.0063  
MAE (testset)     0.7431  0.7303  0.7366  0.7380  0.7350  0.7366  0.0042  
Fit time          0.80    0.83    0.81    0.77    0.77    0.80    0.02    
Test time         0.12    0.10    0.14    0.14    0.15    0.13    0.02    
Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9173  0.9158  0.9211  0.9266  0.9133  0.9188  0.0046  
MAE (testset)     0.7190  0.7194  0.7210  0.7272  0.7164  0.7206  0.0036  
Fit time          12.98   13.39   13.06   13.07   13.88   13.28   0.33    
Test time         3.46    3.53    3.50    3.40    3.55    3.49    0.05    
Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (tests


Проанализуємо результати крос-валідації для кожної моделі.

Модель SVD:
RMSE (середньоквадратична помилка): Середнє значення RMSE для кожного з розгорнутих наборів даних коливається від 0.925 до 0.944 з середнім значенням приблизно 0.934. Стандартне відхилення досить мале - близько 0.0063, що свідчить про стабільність результатів.
MAE (середня абсолютна помилка): Середнє значення MAE знаходиться в діапазоні від 0.730 до 0.743, з середнім значенням близько 0.737. Стандартне відхилення також невелике - приблизно 0.0042.
Модель SVD++:
RMSE: Середнє значення RMSE коливається від 0.914 до 0.927 з середнім значенням приблизно 0.919. Стандартне відхилення досить мале - близько 0.0046, що також свідчить про стабільність результатів.
MAE: Середнє значення MAE знаходиться в діапазоні від 0.716 до 0.727, з середнім значенням близько 0.721. Стандартне відхилення становить близько 0.0036.
Модель NMF:
RMSE: Середнє значення RMSE для кожного з розгорнутих наборів даних коливається від 0.959 до 0.965 з середнім значенням близько 0.962. Стандартне відхилення досить мале - приблизно 0.0020.
MAE: Середнє значення MAE знаходиться в діапазоні від 0.750 до 0.757, з середнім значенням близько 0.755. Стандартне відхилення становить близько 0.0025.
Загальні висновки:
Модель SVD++ має найменше значення RMSE та MAE, що свідчить про кращу точність прогнозування порівняно з іншими моделями.
Модель SVD також демонструє добрі результати, але трохи гірші, ніж у SVD++.
Модель NMF має найбільше значення RMSE та MAE, що може свідчити про меншу точність цієї моделі порівняно з іншими двома.
Загалом, на основі цих результатів можна зробити висновок, що модель SVD++ може бути найкращим варіантом для цього конкретного набору даних з оцінками фільмів.